Dapper.GraphQL.Test/InsertTests.cs

```diff
@@ -33,6 +33,12 @@ public void InsertPerson()
 
                 Assert.True(personId > 0);
 
+                int anotherPersonId = SqlBuilder
+                    .Insert(person)
+                    .ExecuteWithSqlIdentity<Person, int>(db, p => p.Id);
```

Comment:
Let's see if we can also maintain the context of what TEntityType was in the Insert() statement.  That means we'll probably was a second, generic `SqlInsertContext`:

```
public class SqlInsertContext<TEntityType> : SqlInsertContext
{
}
```

Then, we can add a second `ExecuteWithSqlIdentity` method with a single parameter:

```
public static async Task<TIdentityType> ExecuteWithSqlIdentityAsync<TEntityType, TIdentityType>(this SqlInsertContext<TEntityType> context, IDbConnection dbConnection, Func<TEntityType, TIdentityType> identityTypeSelector)
{
    ...
}
```

This will mean that you can still do multiple `Insert` statements, but you can only `ExecuteWithSqlIdentity()` against the final `Insert()` call to automatically infer the generic argument types.  With this in place, the call should look like:

```
int anotherPersonId = SqlBuilder
    .Insert(person)
    .ExecuteWithSqlIdentity(db, p => p.Id);
```

What do you think?  Do you see any issues with this approach?

Reply:
Yea this is exactly what I was thinking and you've highlighted my concern about the multiple inserts being chained. 

So if it's only the final `Insert()` that can automatically infer the type, would we then need to split places like this into separate executions if we wanted the generated `Id`?
```csharp
// Add email and phone number to the person
var insert = SqlBuilder
    .Insert(email)
    .Insert(phone);
var insertedCount = await insert.ExecuteAsync(db);
```

I think that's fine. The limitation is the same as it was before anyway in terms of only being able to get that last INSERT statement's record ID anyway. 

I can see that limitation being a more frustrating later if the gql mutation requests back the nested `email`/`phone` and their `id` field. We'd then be stuck between a perf optimization (run multiple at once) and running individually (to get IDs). Maybe we need a `Tuple<T1, T2, T3>` overload as well that builds a query that can capture and return all the IDs... Or is it always best practise to write your records out to DB then retrieve them (e.g. in case the DB has triggers or does some other manipulation of the data on the way in) rather than just assuming what was sent in the mutation was saved?

Reply:
In that kind of architecture, it's probably better not to use identity or database-generated IDs, and generate IDs before inserting (e.g. shortid, GUID, etc.).

I think this is a fair tradeoff for this scenario.

Reply:
Went ahead and implemented that. I think it's looking good. Tests all passing. Take a look when you've got a sec.